### ISSUE TITLE
Should "^*" be valid?

### ISSUE THREAD

**Issue by ltratt** [rust-lang/regex#639]
At the moment this code:

```rust
use regex::Regex;

fn main() {
    let re = Regex::new("^*$").unwrap();
    println!("{:?}", re.is_match("w"));
}
```

prints `true` when executed. I expected `^*` to be invalid because it's not clear to me what the repetition operator applied to `^` means. Python's regular expression library ("nothing to repeat") and OpenBSD's egrep ("repetition-operator operand invalid") both refuse to compile such a regex. All that said, I'm not sure whether there is a reasonable semantics for `^*` and/or whether the current behaviour is intentional or not. So please feel free to tell me that this is working as expected!

**Comment by BurntSushi** [comment]
One could make a case for disallowing this, sure. It's likely I found allowing it to lead to a simpler implementation. e.g., While Python returns an error on `^*`, it will happily accept `(^)*`, which is a bit strange.

The only hint I have is that the `regex-syntax` docs explicitly call out the fact that repetition operators are allowed to apply to empty sub-expressions:

https://github.com/rust-lang/regex/blob/2a8ddd0b36095d3810f3556627cd59c5014edd01/regex-syntax/src/ast/mod.rs#L154-L158

In terms of what the match semantics are, I think the straight-forward interpretation is fine: `^*` means "match the beginning of the string zero or more times." Now obviously, if `^` matches, then `^^` will also match, and so will `^^^` and so on. (The regex engine has checks in place to avoid executing an infinite loop in cases like these, but those occur even if we disallowed corner cases like this.)

So in your example, `^` does certainly match zero or more times _at any position_, including the position at which `$` also matches. Therefore, the overall regex also matches.

If you change your example to `^+$`, then it _should_ output `false`.

While I think your suggestion is defensible, it 1) makes the implementation more complex unless you give up on consistently reporting errors like Python and 2) would technically be a breaking change at this point. (2) alone would cause me to say "might be worth revisiting for regex 2," but (1) suggests that perhaps the current behavior is fine.

As for precedent, Go's regex engine behaves identically to Rust's: https://play.golang.org/p/nXxbjVizdgx

(I would assume RE2 does as well.)
